### PR TITLE
Add terminate execution API.

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
Enable v8 wasm vm to terminate execution.
Signed-off-by: chaoqin-li1123 <chaoqinli@google.com>

### CODE REVIEW

File: include/proxy-wasm/wasm_vm.h [hunk 1]
```diff
@@ -311,6 +311,7 @@ class WasmVm {
 
   // Integrator operations.
   std::unique_ptr<WasmVmIntegration> &integration() { return integration_; }
+  virtual void terminateExecution() {}
```
Comment (PiotrSikora) [inline include/proxy-wasm/wasm_vm.h:314]: 1) Could you move it higher? Right now, it's between 2 "integration" functions. Perhaps before `isFailed()`?
2) I think using `terminate` as a name should be enough.
3) Please mark it as a pure virtual function.
4) Some documentation would be useful.

File: src/v8/v8.cc [hunk 2]
```diff
@@ -66,6 +68,14 @@ class V8 : public WasmVm {
             const std::unordered_map<uint32_t, std::string> &function_names) override;
   std::string_view getPrecompiledSectionName() override;
   bool link(std::string_view debug_name) override;
+  void terminateExecution() override {
```
Comment (PiotrSikora) [inline src/v8/v8.cc:71]: Could you split declaration and implementation, like other functions in the code?

Also, the placement should match that in the `wasm_vm.h`.

File: src/v8/v8.cc [hunk 3]
```diff
@@ -66,6 +68,14 @@ class V8 : public WasmVm {
...
+    auto *store_impl = reinterpret_cast<wasm::StoreImpl *>(store_.get());
+    auto *isolate = store_impl->isolate();
+    isolate->TerminateExecution();
+    while (isolate->IsExecutionTerminating()) {
+      std::this_thread::yield();
+    }
+  }
```
Comment (PiotrSikora) [inline src/v8/v8.cc:78]: Should we log this event? e.g.
```suggestion
    integration()->trace("[host->vm] Terminated");
  }
```

File: test/runtime_test.cc [hunk 4]
```diff
@@ -245,6 +247,40 @@ TEST_P(TestVM, Callback) {
...
+  ASSERT_TRUE(vm_->load(source, {}, {}));
+
+  TestContext context;
+  vm_->registerCallback(
+      "env", "callback", &callback,
+      &ConvertFunctionWordToUint32<decltype(callback), callback>::convertFunctionWordToUint32);
+
+  vm_->registerCallback(
+      "env", "callback2", &callback2,
+      &ConvertFunctionWordToUint32<decltype(callback2), callback2>::convertFunctionWordToUint32);
```
Comment (PiotrSikora) [inline test/runtime_test.cc:264]: Please create a simple `infinite_loop.wasm` test file.

There is absolutely no reason to add this into `callback.wasm`, especially when those extra callbacks only add to the cognitive complexity ("is this needed for this test or only a result of copy&paste?", etc.).

File: test/runtime_test.cc [hunk 5]
```diff
@@ -245,6 +247,40 @@ TEST_P(TestVM, Callback) {
...
+    std::this_thread::sleep_for(std::chrono::seconds(3));
+    vm_->terminateExecution();
+  });
+
+  ASSERT_TRUE(vm_->link(""));
+  WasmCallWord<1> run2;
+  vm_->getFunction("infinite_loop", &run2);
+  EXPECT_TRUE(run2 != nullptr);
+
+  run2(&context, Word{0});
```
Comment (PiotrSikora) [inline test/runtime_test.cc:276]: This should be named `infinite_loop` (after the function it calls), and it doesn't look that the extra parameter is used for anything.
```suggestion
  WasmCallVoid<0> infinite_loop;
  vm_->getFunction("infinite_loop", &infinite_loop);
  EXPECT_TRUE(infinite_loop != nullptr);

  infinite_loop(&context);
```

File: test/test_data/callback.rs [hunk 6]
```diff
@@ -28,6 +28,23 @@ pub extern "C" fn run2(val: i32) -> i32 {
...
+            count -= 2;
+        }
+        if count >= 100 {
+            break;
+        }
+    }
+    unsafe {
+        callback2(val) + A
+    }
+}
```
Comment (PiotrSikora) [inline test/test_data/callback.rs:46]: Again, there is no reason to make this complex. This simple infinite loop also works:
```suggestion
pub extern "C" fn infinite_loop() {
    let mut _count: u64 = 0;
    loop {
        _count += 1;
    }
}
```

File: include/proxy-wasm/wasm.h [hunk 7]
```diff
@@ -85,7 +85,6 @@ class WasmBase : public std::enable_shared_from_this<WasmBase> {
 
   void timerReady(uint32_t root_context_id);
   void queueReady(uint32_t root_context_id, uint32_t token);
-
```
Comment (PiotrSikora) [inline include/proxy-wasm/wasm.h:88]: Unrelated change, please remvoe it.

File: test/runtime_test.cc [hunk 8]
```diff
@@ -14,11 +14,13 @@
 
 #include "gtest/gtest.h"
 
+#include <cassert>
```
Comment (PiotrSikora) [inline test/runtime_test.cc:17]: This isn't used anywhere.

File: test/runtime_test.cc [hunk 9]
```diff
@@ -245,6 +247,40 @@ TEST_P(TestVM, Callback) {
   ASSERT_EQ(res.u32(), 100100); // 10000 (global) + 100(in callback)
 }
 
+TEST_P(TestVM, TerminateExecution) {
+  if (engine_ != "v8") {
+    return;
```
Comment (PiotrSikora) [inline test/runtime_test.cc:252]: Add note or a `TODO` explaining why this is executed only with V8 and when/if the same functionality is going to be added to other engines.

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: src/v8/v8.cc (hunk 3 above)
Comment (chaoqin-li1123) [reply src/v8/v8.cc:78]: If we don't mark it as failed, the user can may reuse it. If we mark it as fail explicitly, we don't need to wait for the termination to finish.

File: test/runtime_test.cc (hunk 9 above)
Comment (chaoqin-li1123) [reply test/runtime_test.cc:252]: TODO added.

File: src/v8/v8.cc (hunk 3 above)
Comment (chaoqin-li1123) [reply src/v8/v8.cc]: It seems that fail() is not thread safe, but terminate() should be, so I get a tsan error. Do you have any suggestion?

Comment (PiotrSikora) [reply src/v8/v8.cc]: Let's drop explicit `fail()` call for now, since it's going to turn into a trap and failed WasmVM anyway when unrolling.

**Comment by chaoqin-li1123** [comment]
./retest

**Comment by PiotrSikora** [comment]
There is no `/retest` here. I re-kicked the CI, but in the future you can simply push an empty commit to do that yourself.

File: test/test_data/infinite_loop.rs [hunk 10]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2020 Google LLC
```
Comment (PiotrSikora) [inline test/test_data/infinite_loop.rs:1]: Nit: 2022.

File: src/v8/v8.cc [hunk 11]
```diff
@@ -92,6 +94,16 @@ class V8 : public WasmVm {
...
 
+  void terminate() override {
+    auto *store_impl = reinterpret_cast<wasm::StoreImpl *>(store_.get());
+    auto *isolate = store_impl->isolate();
+    isolate->TerminateExecution();
+    while (isolate->IsExecutionTerminating()) {
+      std::this_thread::yield();
+    }
+    integration()->trace("[host->vm] Terminated");
+  }
```
Comment (PiotrSikora) [inline src/v8/v8.cc:105]: As mentioned before, please move implementation outside of the class declaration.

File: include/proxy-wasm/wasm_vm.h [hunk 12]
```diff
@@ -297,6 +297,11 @@ class WasmVm {
   FOR_ALL_WASM_VM_IMPORTS(_REGISTER_CALLBACK)
 #undef _REGISTER_CALLBACK
 
+  /**
+   * Terminate the current execution of wasm vm, the vm fails and shouldn't run after calling
+   * terminate.
+   */
+  virtual void terminate() = 0;
   bool isFailed() { return failed_ != FailState::Ok; }
```
Comment (PiotrSikora) [inline include/proxy-wasm/wasm_vm.h:305]: Nit: add an empty line between `terminate` and `isFailed`.

File: include/proxy-wasm/wasm_vm.h [hunk 13]
```diff
@@ -297,6 +297,11 @@ class WasmVm {
   FOR_ALL_WASM_VM_IMPORTS(_REGISTER_CALLBACK)
 #undef _REGISTER_CALLBACK
 
+  /**
+   * Terminate the current execution of wasm vm, the vm fails and shouldn't run after calling
+   * terminate.
```
Comment (PiotrSikora) [inline include/proxy-wasm/wasm_vm.h:302]: Style: `Terminate execution of this WasmVM. It shouldn't be used after being terminated.`

**Review by PiotrSikora — COMMENTED** [review]
LGTM, with a few style nits.

File: test/runtime_test.cc [hunk 14]
```diff
@@ -245,6 +246,34 @@ TEST_P(TestVM, Callback) {
...
+
+  std::thread terminate([&]() {
+    std::this_thread::sleep_for(std::chrono::seconds(3));
+    vm_->terminate();
+  });
+
+  ASSERT_TRUE(vm_->link(""));
+  WasmCallVoid<0> infinite_loop;
+  vm_->getFunction("infinite_loop", &infinite_loop);
+  EXPECT_TRUE(infinite_loop != nullptr);
```
Comment (PiotrSikora) [inline test/runtime_test.cc:267]: Nit: `ASSERT_TRUE()`.

**Review by PiotrSikora — APPROVED** [review]
LGTM, thanks!
